### DISCUSSION TITLE
IKEv2 uniqueness policy

### DISCUSSION THREAD

**Discussion by bhatnitish** [strongswan/strongswan#1682] [Q&A]
Hi,

We could end up with duplicate IKE SAs with IKEv2 when both sides reauth the IKE SA at the same time, with the default break-before-make behavior. Even with a "keep" uniqueness policy in place, we end up with duplicates over time. I'm guessing this is because the uniqueness check happens before the new SA is established. This won't help as both SAs are created in parallel. Is there a way to enforce a uniqueness check after the SA is established? The uniqueness policy "replace" might help here: "replace deletes any existing connection if a new one for the same user gets established". This sounds like the uniqueness check happens when the new SA to the same user gets ESTABLISHED. Is that true? Is there a robust way to avoid IKE SA duplicates?

**Comment by tobiasbrunner** [comment]
> We could end up with duplicate IKE SAs with IKEv2 when both sides reauth the IKE SA at the same time, with the default break-before-make behavior.

Using different lifetimes (and/or randomization etc.) can maybe avoid that (or you don't use reauthentication at all and rely on rekeying). See [here](https://docs.strongswan.org/docs/5.9/config/rekeying.html) for details.

> Is there a way to enforce a uniqueness check after the SA is established?

No, currently not.

> This sounds like the uniqueness check happens when the new SA to the same user gets ESTABLISHED.

Yes, the uniqueness check happens during the authentication. There are situations where that check is not able to prevent duplicate IKE_SAs. But note that reauthentications are a special case because they are heuristically ignored by that check. Generally, the duplicates shouldn't be an issue.
